File: engine/src/generation.ts
// How an AI writes an element's body, declared by the element.
//
// The ✨ Generate buttons were once five hand-written call sites in the editor,
// which is how `image_view` came to have a `code` field and no way to fill it:
// nobody added the sixth branch. An element declares this instead, so a shell
// renders one button per element and knows nothing about which element it is.
//
// It lives beside `Logic` on purpose. A logic says where the body is kept and
// who runs it; this says how the body gets written in the first place, and both
// name the *same* config keys — passed in as one `LogicFields` constant per
// element, so the two can never drift into naming different fields.

import type { LogicFields } from './logic.ts';

/**
 * Which generator writes the body, and so which prompt it is written with.
 *
 * `code` and `prompt` are the two bodies that run; `data_format` and
 * `output_format` are contracts written for *other* elements to be generated
 * against, which is a different prompt and a different result.
 */
export type GenerationKind = 'code' | 'prompt' | 'output_format' | 'data_format';

/** One element's answer to "how does an AI write this?". */
export interface Generation {
  kind: GenerationKind;
  /** The same constant the element's `logic()` is built from. */
  fields: LogicFields;
  /**
   * What THIS element does with the snippet, in the words the model is given.
   *
   * It lives next to the element that runs the snippet rather than in a shared
   * constants module, because it describes that element's behaviour and stops
   * being true the moment the behaviour changes.
   */
  contract?: string;
  /**
   * A sub-snippet's fixed port names. Omitted means "whatever this node is
   * actually wired as", which is what a code node wants — a transform inside a
   * block, by contrast, always receives `value` and returns `value`.
   */
  inputs?: string[];
  outputs?: string[];
  /** Shown when the request field is still empty. */
  guard: string;
  /** Shown when the generated text arrives. */
  success: string;
}

/**
 * The directory-mode file selector, declared once because it is implemented
 * once: an input node and a file-picker block run the identical
 * `run(inputs) -> {files}` contract, one as a node and one as a block inside a
 * page. The node/widget split is not a behavioural difference here, and two
 * copies of a prompt drift.
 */
export const SELECTOR_FIELDS: LogicFields = {
  body: 'selector_code', prompt: 'selector_prompt', file: 'code_file',
};

export const SELECTOR_GENERATION: Generation = {
  kind: 'code',
  fields: SELECTOR_FIELDS,
  contract:
    '`inputs["files"]` is the full list of rooted file paths found in the '
    + 'directory. Return only the selected paths as {"files": [...]}.',
  inputs: ['files'],
  outputs: ['files'],
  guard: 'Please describe which files to select first.',
  success: '✅ Selector generated!',
};

File: backend/app/services/generation.py
"""
How an element's body gets written by an AI, asked of the engine.

The elements live in the engine, in one language, and they declare this next to
the `logic()` that says where the body is kept -- both built from one constant
per element, so the button cannot write into a field nothing runs. This module
is only the shape that answer arrives in on this side.

Cached for the life of the process: a generation descriptor is a property of the
element, not of a node, so it does not change while the engine is up.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional, Tuple

from app.services import engine_client


@dataclass(frozen=True)
class Generation:
    """One element's answer, flattened. Field names follow the wire."""

    kind: str                                  # "code" | "prompt" | "output_format" | "data_format"
    target_field: str                          # where the generated text is written
    prompt_field: str                          # where the request is stored
    prompt_on_node: bool = False               # the ai node's request is node.description
    contract: str = ""
    # Set only for a sub-snippet whose ports the element fixes (a selector's
    # `files`, a transform's `value`). None means "the node's real ports", which
    # is what a code node wants.
    inputs: Optional[Tuple[str, ...]] = None
    outputs: Optional[Tuple[str, ...]] = None
    guard: str = ""
    success: str = ""

    @classmethod
    def from_engine(cls, row: Dict[str, Any]) -> "Generation":
        return cls(
            kind=str(row["kind"]),
            target_field=str(row["target_field"]),
            prompt_field=str(row["prompt_field"]),
            prompt_on_node=bool(row.get("prompt_on_node")),
            contract=str(row.get("contract") or ""),
            inputs=tuple(row["inputs"]) if row.get("inputs") else None,
            outputs=tuple(row["outputs"]) if row.get("outputs") else None,
            guard=str(row.get("guard") or ""),
            success=str(row.get("success") or ""),
        )


_cache: Optional[Dict[str, Generation]] = None


async def all_generations() -> Dict[str, Generation]:
    """Every element that can have its body written for it, by name."""
    global _cache
    if _cache is None:
        raw = await engine_client.get("/api/elements/generation")
        _cache = {name: Generation.from_engine(row) for name, row in raw.items()}
    return _cache


async def generation_for(element: str) -> Optional[Generation]:
    """The descriptor for one element, or None if it authors nothing.

    Node types and block kinds share one namespace, because a caller has one
    name and no reason to know which level it came from.
    """
    return (await all_generations()).get(element)


def forget() -> None:
    """Drop the cache. For tests that restart the engine under it."""
    global _cache
    _cache = None
